File: python/src/falcon_core/math/discrete_spaces/discrete_space.py
from __future__ import annotations
from typing import Any, List, Dict, Tuple, Optional
from falcon_core._capi.discrete_space import DiscreteSpace as _CDiscreteSpace
from falcon_core.math.axes import Axes
from falcon_core.math.axes import Axes
from falcon_core.math.axes import Axes
from falcon_core.math.axes import Axes
from falcon_core.math.axes import Axes
from falcon_core.math.domains.coupled_labelled_domain import CoupledLabelledDomain
from falcon_core.math.domains.domain import Domain
from falcon_core.instrument_interfaces.names.instrument_port import InstrumentPort
from falcon_core.generic.map import Map
from falcon_core.instrument_interfaces.names.ports import Ports
from falcon_core.math.unit_space import UnitSpace

class DiscreteSpace:
    """Python wrapper for DiscreteSpace."""

    def __init__(self, c_obj):
        self._c = c_obj

    @classmethod
    def _from_capi(cls, c_obj):
        if c_obj is None:
            return None
        return cls(c_obj)

    @classmethod
    def new(cls, space: UnitSpace, axes: Axes, increasing: Axes) -> DiscreteSpace:
        return cls(_CDiscreteSpace.new(space._c if space is not None else None, axes._c if axes is not None else None, increasing._c if increasing is not None else None))

    @classmethod
    def new_cartesian_discrete_space(cls, divisions: Axes, axes: Axes, increasing: Axes, domain: Domain) -> DiscreteSpace:
        return cls(_CDiscreteSpace.new_cartesian_discrete_space(divisions._c if divisions is not None else None, axes._c if axes is not None else None, increasing._c if increasing is not None else None, domain._c if domain is not None else None))

    @classmethod
    def new_cartesian_discrete_space_1D(cls, division: Any, shared_domain: CoupledLabelledDomain, increasing: Map, domain: Domain) -> DiscreteSpace:
        return cls(_CDiscreteSpace.new_cartesian_discrete_space_1D(division, shared_domain._c if shared_domain is not None else None, increasing._c if increasing is not None else None, domain._c if domain is not None else None))

    @classmethod
    def from_json(cls, json: str) -> DiscreteSpace:
        return cls(_CDiscreteSpace.from_json(json))

    def space(self, ) -> UnitSpace:
        ret = self._c.space()
        if ret is None: return None
        return UnitSpace._from_capi(ret)

    def axes(self, ) -> Axes:
        ret = self._c.axes()
        if ret is None: return None
        return Axes(ret)

    def increasing(self, ) -> Axes:
        ret = self._c.increasing()
        if ret is None: return None
        return Axes(ret)

    def knobs(self, ) -> Ports:
        ret = self._c.knobs()
        if ret is None: return None
        return Ports._from_capi(ret)

    def validate_unit_space_dimensionality_matches_knobs(self, ) -> None:
        ret = self._c.validate_unit_space_dimensionality_matches_knobs()
        return ret

    def validate_knob_uniqueness(self, ) -> None:
        ret = self._c.validate_knob_uniqueness()
        return ret

    def get_axis(self, knob: InstrumentPort) -> None:
        ret = self._c.get_axis(knob._c if knob is not None else None)
        return ret

    def get_domain(self, knob: InstrumentPort) -> Domain:
        ret = self._c.get_domain(knob._c if knob is not None else None)
        if ret is None: return None
        return Domain._from_capi(ret)

    def get_projection(self, projection: Axes) -> Axes:
        ret = self._c.get_projection(projection._c if projection is not None else None)
        if ret is None: return None
        return Axes(ret)

    def equal(self, other: DiscreteSpace) -> None:
        ret = self._c.equal(other._c if other is not None else None)
        return ret

    def not_equal(self, other: DiscreteSpace) -> None:
        ret = self._c.not_equal(other._c if other is not None else None)
        return ret

    def to_json(self, ) -> str:
        ret = self._c.to_json()
        return ret

    def __eq__(self, other):
        """Operator overload for =="""
        if not isinstance(other, DiscreteSpace):
            return NotImplemented
        return self.equal(other)

    def __ne__(self, other):
        """Operator overload for !="""
        if not isinstance(other, DiscreteSpace):
            return NotImplemented
        return self.not_equal(other)

File: ocaml/_opam/.opam-switch/sources/falcon_core/src/falcon_core/math/discrete_spaces/discretespace.ml
open Ctypes
open Capi_bindings
open ErrorHandling

open Falcon_core.Generic
open Falcon_core.Instrument_interfaces.Names
open Falcon_core.Math
open Falcon_core.Math.Domains

class c_discretespace (h : unit ptr) = object(self)
  val raw_val = h
  method raw = raw_val
  initializer Gc.finalise (fun _ ->
    Capi_bindings.discretespace_destroy raw_val;
    ErrorHandling.raise_if_error ()
  ) self
end

module DiscreteSpace = struct
  type t = c_discretespace

  let copy (handle : t) : t =
    ErrorHandling.read handle (fun () ->
      let ptr = Capi_bindings.discretespace_copy handle#raw in
      ErrorHandling.raise_if_error ();
      new c_discretespace ptr
    )

  let fromjson (json : string) : t =
    let ptr = Capi_bindings.discretespace_from_json_string (Capi_bindings.string_wrap json) in
    ErrorHandling.raise_if_error ();
    new c_discretespace ptr

  let make (space : UnitSpace.t) (axes : AxesCoupledLabelledDomain.t) (increasing : AxesMapStringBool.t) : t =
    ErrorHandling.multi_read [space; axes; increasing] (fun () ->
      let ptr = Capi_bindings.discretespace_create space#raw axes#raw increasing#raw in
      ErrorHandling.raise_if_error ();
      new c_discretespace ptr
    )

  let cartesianDiscreteSpace (divisions : AxesInt.t) (axes : AxesCoupledLabelledDomain.t) (increasing : AxesMapStringBool.t) (domain : Domain.t) : t =
    ErrorHandling.multi_read [divisions; axes; increasing; domain] (fun () ->
      let ptr = Capi_bindings.discretespace_create_cartesian_discrete_space divisions#raw axes#raw increasing#raw domain#raw in
      ErrorHandling.raise_if_error ();
      new c_discretespace ptr
    )

  let cartesianDiscreteSpace1d (division : int) (shared_domain : CoupledLabelledDomain.t) (increasing : MapStringBool.t) (domain : Domain.t) : t =
    ErrorHandling.multi_read [shared_domain; increasing; domain] (fun () ->
      let ptr = Capi_bindings.discretespace_create_cartesian_discrete_space_1d division shared_domain#raw increasing#raw domain#raw in
      ErrorHandling.raise_if_error ();
      new c_discretespace ptr
    )

  let equal (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.discretespace_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let notEqual (handle : t) (other : t) : bool =
    ErrorHandling.multi_read [handle; other] (fun () ->
      let result = Capi_bindings.discretespace_not_equal handle#raw other#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let toJsonString (handle : t) : string =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_to_json_string handle#raw in
      ErrorHandling.raise_if_error ();
      Capi_bindings.string_to_ocaml result
    )

  let space (handle : t) : UnitSpace.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_space handle#raw in
      ErrorHandling.raise_if_error ();
      new c_unitspace result
    )

  let axes (handle : t) : AxesCoupledLabelledDomain.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_axes handle#raw in
      ErrorHandling.raise_if_error ();
      new c_axescoupledlabelleddomain result
    )

  let increasing (handle : t) : AxesMapStringBool.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_increasing handle#raw in
      ErrorHandling.raise_if_error ();
      new c_axesmapstringbool result
    )

  let knobs (handle : t) : Ports.t =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_knobs handle#raw in
      ErrorHandling.raise_if_error ();
      new c_ports result
    )

  let validateUnitSpaceDimensionalityMatchesKnobs (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_validate_unit_space_dimensionality_matches_knobs handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let validateKnobUniqueness (handle : t) : unit =
    ErrorHandling.read handle (fun () ->
      let result = Capi_bindings.discretespace_validate_knob_uniqueness handle#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let getAxis (handle : t) (knob : InstrumentPort.t) : int =
    ErrorHandling.multi_read [handle; knob] (fun () ->
      let result = Capi_bindings.discretespace_get_axis handle#raw knob#raw in
      ErrorHandling.raise_if_error ();
      result
    )

  let getDomain (handle : t) (knob : InstrumentPort.t) : Domain.t =
    ErrorHandling.multi_read [handle; knob] (fun () ->
      let result = Capi_bindings.discretespace_get_domain handle#raw knob#raw in
      ErrorHandling.raise_if_error ();
      new c_domain result
    )

  let getProjection (handle : t) (projection : AxesInstrumentPort.t) : AxesLabelledControlArray.t =
    ErrorHandling.multi_read [handle; projection] (fun () ->
      let result = Capi_bindings.discretespace_get_projection handle#raw projection#raw in
      ErrorHandling.raise_if_error ();
      new c_axeslabelledcontrolarray result
    )

end